### PR TITLE
Interactive shell implementation

### LINKED ISSUE #329: Interactivity
Make LPython work in an interactive prompt.

Copy how LFortran does it. We have to remember a symbol table, and compile each prompt as we go. LFortran has all the infrastructure, I think it's not difficult to add to LPython.

### PR DESCRIPTION
**Pull request by Vipul-Cariappa**
Fixes #329

# TODOs
- [ ] Remember global functions and variables declared.
- [ ] Error message with proper line number and highlighting
- [ ] Intrinsic functions support
- [ ] print the value of the evaluated expression
- [ ] Import other modules
- [ ] `symengine` and `numpy` support
- [ ] Support complex data types like `list`, `tuple` and `dict`

Got the following things to work:
image

### CODE REVIEW

**Comment by Vipul-Cariappa** [comment]
I want to know how should I proceed with the first to-do point "Remember global functions and variables declared". The problem is as follows

```bash
>>> def f():
...   print("called function f")
... 
>>> f()
semantic error: Function 'f' is not declared and not intrinsic
 --> input:1:1
  |
1 | f()
  | ^ 
```

Another example:
```bash
>>> x: i32 = 10
>>> print(x)
semantic error: Variable 'x' not declared
 --> input:1:1
  |
1 | print(x)
  | ^ 
```

The declared function `f` is lost by the time we evaluate (i.e. parse and execute) the function call to `f`. We should store its declaration (and `x`'s declaration) somewhere (I believe at the AST level) and include it while parsing the second part/prompt.

I want to know if there are existing methods to do this. Or should I implement it myself. I want something like.
`get_all_function_declerations(AST_t*)` and `get_all_global_declerations(AST_t*)`, and `extend_ast_with_function_decleration(AST_t*, Function_t*)` and `extend_ast_with_global_decleration(AST_t*, ...)`

**Comment by Shaikh-Ubaid** [comment]
I think the idea to solve this is to keep updating the original ASR on every new interactive query run by the user.

On every run, the functions or variables defined in the previous run need to be updated. We remove the body of functions and set their `abiType` to interactive. Similarly for variables, we mark them as interactive. Later we handle the `interactive` functions/variables appropriately in the `AST->ASR`, `ASR passes` and `ASR->backend`. 

LFortran already supports interactive compilation. You can look into it for reference. Since the ASR passes and backends are shared with LFortran, I think we will mostly need interactive support for AST->ASR in LPython.

**Comment by Shaikh-Ubaid** [comment]
Let's build this iteratively. For this PR, let's get something that works with the interactive shell, add tests for the implemented support and ensure the CI passes. After that, you can send more PRs to improve the existing support and/or add more features as planned in https://github.com/lcompilers/lpython/pull/2617#issue-2195119178. Ideally, small and frequent PRs that just focus on one particular support with robust tests are preferable.

For tests, you can look into how lfortran does it https://github.com/lfortran/lfortran/blob/582db3d4379eee26cfeb0721d5fda21971aaae3d/src/lfortran/tests/test_llvm.cpp#L439-L455.

**Comment by Vipul-Cariappa** [comment]
It makes more sense to have multiple smaller PR than a large one. I will finish up the work with remembering context from the previous cells, i.e. remembering global functions and variables.
We can add in tests and merge this PR. The other points can be their own separate PRs.

File: src/lpython/python_evaluator.cpp
Comment (Vipul-Cariappa) [inline src/lpython/python_evaluator.cpp:203]: The argument `""` is the `run_fn`. The `run_fn` is not used here. In LFortran all global statements are compiled into 1 function, but in LPython it is compiled into two different functions one for global initialization and the other to execute global statements.

File: src/lpython/semantics/python_ast_to_asr.cpp
Comment (Vipul-Cariappa) [inline src/lpython/semantics/python_ast_to_asr.cpp:5039]: Any suggestions on this?
Should we create `run_fn1` and `run_fn2` and share it through the `PassOptions`?
But even that may not be straightforward.

**Comment by Vipul-Cariappa** [comment]
I guess we can merge the work done till this point. The redefinition of symbols is not yet supported. It looks more complicated than initially thought. We can add support for the redefinition of symbols in a later PR.

Not sure why the tests are failing.

I have left 2 comments please have a look at it.

File: src/lpython/semantics/python_ast_to_asr.cpp
Comment (Vipul-Cariappa) [reply src/lpython/semantics/python_ast_to_asr.cpp:5039]: Another option would be to combine the `init` and `stmts` functions. Look at #2573

**Comment by Shaikh-Ubaid** [comment]
You need to update the reference tests. You can do `./run_tests.py -u`.

File: src/lpython/semantics/python_ast_to_asr.cpp
Comment (Shaikh-Ubaid) [reply src/lpython/semantics/python_ast_to_asr.cpp:5039]: I saw the linked issue. If combining `init` and `stmt` function solves your issue, then you can try it. Submit it as a separate PR and we can rebase this after that is merged.

**Comment by Shaikh-Ubaid** [comment]
At a high level, it looks good. I will review in detail in evening today.

File: src/lpython/parser/parser.cpp
Comment (Shaikh-Ubaid) [inline src/lpython/parser/parser.cpp:69]: Any specific reason why we use `make_Interactive_t` here? It seems if we reuse the existing `parse()` function (on line `17` of this file) which returns a `Module` we can save a lot of code being added in this PR.

For example, almost all the code added in `python_ast_to_asr.cpp` would not be needed. It seems to be duplicate of existing code for `visit_Module()`.

File: src/lpython/semantics/python_ast_to_asr.h
Comment (Shaikh-Ubaid) [inline src/lpython/semantics/python_ast_to_asr.h:11]: You can also pass the `eval_count` here and have it as a class member. Then for `body visitor` inside `visit_Module` (assuming we are reusing this instead of defining `visit_Interactive`), you can use the `eval_count` for naming `global_init` and `global_stmts` if `interactive = true`. (For example simply append the eval_count to the end if interactive is true).

File: src/lpython/semantics/python_ast_to_asr.cpp
Comment (Shaikh-Ubaid) [inline src/lpython/semantics/python_ast_to_asr.cpp:8580]: Do we need this here? Is this a work in progress?

File: src/lpython/python_evaluator.cpp
Comment (Shaikh-Ubaid) [inline src/lpython/python_evaluator.cpp:104]: Construct this like

```
std::string base_fn_name = "_main_____main____lpython_interactive";
std::string init_fn = base_fn_name + "_init_" + std::to_string(eval_count);
std::string module_init_fn = "__module__" + init_fn;
```

Similarly for `stmts`. Now this can be reused when erasing symbol below.

File: src/lpython/python_evaluator.cpp
Comment (Shaikh-Ubaid) [inline src/lpython/python_evaluator.cpp:127]: Reuse the variables from the suggestion above.

**Review by Shaikh-Ubaid — COMMENTED** [review]
I shared a couple comments. 

Mainly:
- Let's reuse `Module`, that way we get to reuse a lot of other existing code
- We can merge `global_init` and `global_stmts` in a separate PR. I think we need it to solve https://github.com/lcompilers/lpython/issues/2573
- Minor refactoring

Apart from that, it looks good. Great start! Amazing work!

File: src/bin/lpython.cpp
Comment (Shaikh-Ubaid) [inline src/bin/lpython.cpp:1946]: Do you need this change for `interactivity`? 

It should not be here. `lpython_pass_manager.use_default_passes();` applies ASR passes to the output of `--show-asr`. This changes the reference results and therefore the failure at the CI.

Let's revert this back.

**Comment by Shaikh-Ubaid** [comment]
Please mark as "Ready for review" when ready.

File: src/lpython/python_evaluator.cpp
Comment (Vipul-Cariappa) [reply src/lpython/python_evaluator.cpp:203]: Now that we have merged `global_init` and `global_stmts`, we can assign `run_fn` to `global_stmts` and use it.

Then that change would be part of this PR. You can have a look at it in successive commits.

File: src/bin/lpython.cpp
Comment (Vipul-Cariappa) [reply src/bin/lpython.cpp:1946]: `lpython_pass_manager.parse_pass_arg(arg_pass, skip_pass);` and `lpython_pass_manager.use_default_passes();` is required to be initilised before calling `interactive_python_repl`. I will remove it from here and put it just before calling `interactive_python_repl`.

Overall behavior will be the same as before.

File: src/lpython/semantics/python_ast_to_asr.cpp
Comment (Vipul-Cariappa) [reply src/lpython/semantics/python_ast_to_asr.cpp:8580]: This is not required. There are other changes that givens the same effect as creating a `SymbolTable` here.

File: src/lpython/parser/parser.cpp
Comment (Vipul-Cariappa) [reply src/lpython/parser/parser.cpp:69]: There are some differences between `SymbolTreeVisitor.visit_Interactive` and `SymbolTreeVisitor.visit_Module`.
I will try merging both. Most probably it should be possible. Then we can remove `make_Interactive_t` and reuse `visit_Module` at the places `visit_Interactive` is used.

File: src/lpython/python_evaluator.cpp
Comment (Shaikh-Ubaid) [inline src/lpython/python_evaluator.cpp:12]: I think this should be included with other system headers at the top. I am unsure if system headers should be listed first and then the custom headers or the other way round. But I think we should atleast not mix them.

File: src/lpython/python_evaluator.cpp
Comment (Shaikh-Ubaid) [inline src/lpython/python_evaluator.cpp:52]: We later plan to move the interactivity support into the LibASR so that it is shared between LFortran, LPython and LC. So, it is better that we have/use as common/generic code as possible which easily gets shared across all LCompilers.

Assuming it does not affect the newly added interactivity, I think for now it is better to have it.

File: src/lpython/python_evaluator.h
Comment (Shaikh-Ubaid) [inline src/lpython/python_evaluator.h:74]: This is passed as argument to `python_ast_to_asr()`. The call to `python_ast_to_asr()` is also used by other backends. So, this should be declared outside of `#ifdef HAVE_LFORTRAN_LLVM`.

**Review by Shaikh-Ubaid — COMMENTED** [review]
Good! Few more minor comments. I think it will be good to merge after that. We can add tests in a separate PR.

File: src/lpython/python_evaluator.cpp
Comment (Vipul-Cariappa) [reply src/lpython/python_evaluator.cpp:52]: Can we keep this as part of the next PR? It may require changes in more than one place.

> I was working towards "print the value of the evaluated expression" (https://github.com/lcompilers/lpython/pull/2617#issue-2195119178) goal, and I think there might be some obstacles to doing this. But is doable.

**Comment by Vipul-Cariappa** [comment]
I have made the suggested changes. Please have a look. I have also left a comment on one of your suggestions.

File: src/lpython/python_evaluator.cpp
Comment (Shaikh-Ubaid) [reply src/lpython/python_evaluator.cpp:52]: >Can we keep this as part of the next PR?

We can do that.

**Review by Shaikh-Ubaid — APPROVED** [review]
It looks good! Thanks for this! Great work!

**Comment by Shaikh-Ubaid** [comment]
Let's add few/some tests for this in a next PR!

**Pull request merged**
